src/shared/fdset.c

```diff
@@ -97,21 +97,33 @@ int fdset_consume(FDSet *s, int fd) {
         return r;
 }
 
-int fdset_put_dup(FDSet *s, int fd) {
+int fdset_put_dup_full(FDSet *s, int fd, bool store_index) {
         _cleanup_close_ int copy = -EBADF;
+        unsigned min_fd = 3;
         int r;
 
         assert(s);
         assert(fd >= 0);
 
-        copy = fcntl(fd, F_DUPFD_CLOEXEC, 3);
+        /* If we store by index, then we need to ensure new FDs are always increasing monotonically,
+         * otherwise if we jump back the assumption is going to break. This is used when serializing
```

Comment:
"the assumption"? which assumption?

not grokking the comment

Reply:
that the newly added FD will be at the end of the set, which is what makes this scheme you suggested work - I'll amend the comment